### DISCUSSION TITLE
Expand range of configurations tested on CI ("latest" and "defaults/lowerbound")

### DISCUSSION THREAD

**Discussion by kelset** [react-native-community/discussions-and-proposals#687] [Ideas]
Opening this conversation to collect some ideas about a problem we've hit a couple times recently in the 0.72 release cycle but that has been [around since forever](https://github.com/react-native-community/discussions-and-proposals/issues/17#issuecomment-446261269): there have been a couple of occasions in which changes have been introduced (in main) that would only be compatible with Xcode 14.3, forcing developers to update their toolchain if they wanted to use these new patch releases.

Namely:
* https://github.com/facebook/react-native/commit/8b1bf058c4bcbf4e5ca45b0056217266a1ed870c this one never made it into a patch release 'cause [we realised in time](https://github.com/reactwg/react-native-releases/discussions/73#discussioncomment-6195727) that would force consumers to upgrade to Xcode 14.3
* https://github.com/facebook/react-native/commit/a4a065549677c61eb91bf587032976ed48c75821; we didn't catch this in time and it created effectively a breaking change in a patch release (0.72.2) -> https://github.com/facebook/react-native/issues/38294 that we had to revert via 0.72.3.

In these specific scenarios, why is this considered a blocker if Xcode 14.3.1 is the latest stable release of the tool? Two data points:
* some CIs like GitHub Actions do still default to [Xcode 14.2](https://github.com/actions/runner-images/blob/main/images/macos/macos-12-Readme.md#xcode) in [their images](https://github.com/actions/runner-images)
* Apple's policy is that you need to use [Xcode 14.1](https://developer.apple.com/news/upcoming-requirements/?id=04252023a) or higher

So introducing those changes is an arbitrary breaking change that would disrupt VALID developers' workflows and product pipelines - which is not good.

In terms of compatibility tables, on the Android side of the world the AGP version creates the lowerbound of the Android Studio version that is supported, as described here [developer.android.com/studio/releases#api-level-support](https://developer.android.com/studio/releases#api-level-support) (which makes it easier to track what tools will work with what version) and the [Google Play requirements](https://developer.android.com/google/play/requirements/target-sdk) set the lower boundary for target API, for example right now from August 31st is going to be "New apps must target Android 13 (API level 33) or higher; except for Wear OS apps, which must target Android 11 (API level 30)".

It would be great if we could figure out a way to honour the lowerbound accepted configurations without necessarily forcing React Native itself to stick to these "older" versions. Or, to say in another way: how can we ensure that situations like https://github.com/facebook/react-native/issues/38294 don't go undetected and get flagged only when developers hit the problem (which then force a rapid response to address)?

Me and @fortmarek have been discussing this and we were wondering if a good way to create a first check would be to expand the CI to test both what is being currently used ("latest") AND this lowerbound set up ("defaults" or "lowerbound"). We do have [something in place](https://github.com/facebook/react-native/blob/main/.circleci/config.yml#L102) for testing Node's current LTS and the previous (`nodelts` and `nodeprevlts`) and Ruby (with versions [2.7.7](https://github.com/facebook/react-native/blob/main/.circleci/config.yml#L1664) and [3.2.0](https://github.com/facebook/react-native/blob/main/.circleci/config.yml#L1674)).

For example, if we had in place a similar thing for [Xcode's CI config](https://github.com/facebook/react-native/blob/0.72-stable/.circleci/config.yml#L49), those changes breaking in versions lower than 14.3 would have been detected.

These "compatibility checks" would need to be verified right before every branch cut - so that there is a clear signal of what a version of RN is compatible/expected to be compatible with (and so not to make the versions go stale/forgotten).

Thoughts? Opinions? Leave them below!

**Comment by liamjones** [comment]
I haven't really got anything to add beyond saying, from a purely selfish point of view as an RN user, this sounds like a good idea to me.

From past experience with RN and native iOS development, forced Xcode upgrades for developers can be problematic as;
1. They are sometimes tied to a macOS upgrade as well. That can then result in other unforeseen issues completely unrelated to Xcode. Especially if the new OS version hasn't gone through many patches to get the bugs out yet.
2. They sometimes dictate a Swift upgrade too, requiring various alterations to code due to language changes (though I've not really hit issues with this in recent years while working on RN projects).
3. Upgrading Xcode also implicitly upgrades the resulting iOS/etc SDKs that are used for builds. https://xcodereleases.com/ lists the Xcode/SDK version associations. Sometimes, rarely from my experience, this breaks things. For example, we are currently forced to build our RN projects on Xcode 14.2 or earlier as, as soon as we move to the iOS 16.4 (20E5212f) SDK or higher we lose all inline HTML styling from WKWebViews when using UIPrintPageRenderer.

So the longer RN can support Xcode versions that are still valid for generating builds for the App Store the better, in my opinion. 😄

**Reply by cortinico** [reply]
> So the longer RN can support Xcode versions that are still valid for generating builds for the App Store the better, in my opinion. 😄

This is a valid point, but clashes with the reality of facts that we do have limited resources (in terms on CI time and human time), and we can't test every possible combination.

The same problem applies to Android, Java, Kotlin, Gradle, Node versions and so on.

**Comment by cortinico** [comment]
> It would be great if we could figure out a way to honour the lowerbound accepted configurations without necessarily forcing React Native itself to stick to these "older" versions

Yup I think we should define rangers for each of the tool we support and have a CI job that tests lowerbound + default.

**Reply by kelset** [reply]
I think we should also think about which tool we need to support; for example, I don't think there's much value in checking both Node LTS and the previous LTS.

**Reply by cortinico** [reply]
Yup for Node, we probably need just the latest LTS + the current stable (so in this case 18 and 20)

**Comment by lunaleaps** [comment]
Yea I'm all for adding another configuration with the lower-bound Xcode tested. Would this have caught this issue though, was it just a 14.2 issue? Is it too costly to test all minor versions in between for building?

**Comment by un-variable** [comment]
Expanding the CI configurations to include testing with both the "**latest**" and "**defaults/lowerbound**" versions of Xcode is a practical idea.
 This would help detect compatibility issues early on and provide clarity on which Xcode versions are expected to work with React Native. 
It aligns with the approach taken in other CI setups for Node and Ruby to ensure smoother developer workflows and product pipelines. 
#### This proactive approach can prevent disruptions and improve the overall developer experience.

Certainly, here's a conceptual code snippet in Python for expanding CI configurations to test different Xcode versions:

`import os`

          def test_xcode_compatibility(xcode_version):
         # Configure CI to use the specified Xcode version
          configure_ci(xcode_version)

         # Run tests and checks for React Native compatibility
          run_tests_and_checks()

          def configure_ci(xcode_version):
          If xcode_version == "latest":
          os.environ["XCODE_VERSION"] = "14.3"  # Set the latest Xcode version
          elif xcode_version == "defaults":
          os.environ["XCODE_VERSION"] = "14.2"  # Set default Xcode version
           elif xcode_version == "lowerbound":
          os.environ["XCODE_VERSION"] = "12.0"  # Set lowerbound Xcode version`

          def run_tests_and_checks():
          # Execute tests and checks for React Native compatibility
          print(f"Running tests and checks for Xcode {os.environ['XCODE_VERSION']}...")`

           # Test with the latest Xcode version
           test_xcode_compatibility("latest")

           # Test with the default Xcode version
           test_xcode_compatibility("defaults")

            # Test with a lowerbound Xcode version
            String test_xcode_compatibility("lowerbound")

This Python script demonstrates how you might configure your CI to test React Native compatibility with different Xcode versions, including the "latest," "defaults," and "lowerbound" options. Depending on the specified version, it sets the appropriate Xcode environment variable and runs tests and checks accordingly.

**Comment by cipolleschi** [comment]
I tried to change the executor for a couple of jobs to Xcode 14.1, in order to have a lowerbound check: https://github.com/facebook/react-native/pull/39602

It turns out, the executor which Xcode 14.1 comes with `node 16`... and we need `node 18` for React Native.
I'm looking into how to install a new version of node on CircleCI
